File: Code/Native/Source/LightSystem/RPSpotLight.cpp

#include "RPSpotLight.h"

#define _USE_MATH_DEFINES
#include <math.h>


/**
 * @brief Creates a new spot light
 * @details This creates a new spot light with default properties set. You should
 *   set at least a direction, fov, radius and position to make the light useful.
 */
RPSpotLight::RPSpotLight() : RPLight(RPLight::LT_spot_light) {
    _radius = 10.0;
    _fov = 45.0;
    _direction.set(0, 0, -1);
}

/**
 * @brief Writes the light to a GPUCommand
 * @details This writes the spot light data to a GPUCommand.
 * @see RPLight::write_to_command
 * 
 * @param cmd The target GPUCommand
 */
void RPSpotLight::write_to_command(GPUCommand &cmd) {
    RPLight::write_to_command(cmd);
    cmd.push_float(_radius);

    // Encode FOV as cos(fov)
    cmd.push_float(cos(_fov / 360.0 * M_PI));
    cmd.push_vec3(_direction);
}

/**
 * @brief Inits the shadow sources of the light
 * @details This inits all required shadow sources for the spot light.
 * @see RPLight::init_shadow_sources
 */
void RPSpotLight::init_shadow_sources() {
    nassertv(_shadow_sources.size() == 0);
    _shadow_sources.push_back(new ShadowSource());
}

/**
 * @brief Updates the shadow sources
 * @details This updates all shadow sources of the light.
 * @see RPLight::update_shadow_sources
 */
void RPSpotLight::update_shadow_sources() {
    _shadow_sources[0]->set_resolution(get_shadow_map_resolution());
    _shadow_sources[0]->set_perspective_lens(_fov, _near_plane, _radius, _position, _direction);
}

File: Code/Native/PythonImpl/RPSpotLight.py

import math
from panda3d.core import Vec3

from ShadowSource import ShadowSource

class RPSpotLight(object):

    def __init__(self):
        RPLight.__init__(self, RPLight.LT_spot_light)
        self._radius = 10.0
        self._fov = 45.0
        self._direction = Vec3(0, 0, -1)

    def write_to_command(self, cmd):
        RPLight.write_to_command(self, cmd)
        cmd.push_float(self._radius)
        cmd.push_float(math.cos(self._fov / 360.0 * math.pi))
        cmd.push_vec3(self._direction)

    def set_radius(self, radius):
        self._radius = radius
        self.set_needs_update(True)
        self.invalidate_shadows()

    def get_radius(self):
        return self._radius

    radius = property(get_radius, set_radius)

    def set_fov(self, fov):
        self._fov = fov
        self.set_needs_update(True)
        self.invalidate_shadows()

    def get_fov(self):
        return self._fov

    fov = property(get_fov, set_fov)

    def set_direction(self, *args):
        self._direction = Vec3(*args)
        self._direction.normalize()
        self.set_needs_update(True)
        self.invalidate_shadows()

    def get_direction(self):
        return self._direction

    direction = property(get_direction, set_direction)

    def look_at(self, *args):
        vec = Vec3(*args) - self._position
        self.set_direction(vec)

    def init_shadow_sources(self):
        self._shadow_sources.append(ShadowSource())

    def update_shadow_sources(self):
        self._shadow_sources[0].set_resolution(self.get_shadow_map_resolution())
        self._shadow_sources[0].set_perspective_lens(
            self._fov, self._near_plane, self._radius, self._position, self._direction)
